docs/lang/memories-by-reference.md

```diff
@@ -15,13 +15,58 @@ int main() {
   std::array<uint32_t, 1> x = { 0 };
   add_one(x); // The value at x[0] is now 1.
 }
-``` 
+```
+
+There are two steps to passing a memory by reference in Calyx:
+1. Define the component in a manner such that it can accept a memory by reference.
+2. Pass the desired memory by reference.
+
+The language provides two ways to doing this.
+
+## The Easy Way
+
+Calyx uses the `ref` keyword to describe cells that are passed in by-reference:
+
+```
+component add_one() -> () {
+  cells {
+    ref mem = std_mem_d1(32, 4, 3); // A memory passed in by refernece.
+    ...
+  }
+  ...
+}
+```
+
+This component define `mem` as a memory that is passed in by reference to the component.
+Inside the component we can use the cell like any other cell in the program.
+
+Next, to pass the memory to the component, we can use the `invoke` syntax:
+```
+component add_one() -> () { ... }
+component main() -> () {
+  cells {
+    A = std_mem_d1(32, 4, 3); // A memory passed in by refernece.
```

Comment:
The comment does not apply here. Maybe we should point out that this is an ordinary owned subcomponent instead?